qualtran/_infra/data_types.py

```diff
@@ -420,3 +420,49 @@ def assert_valid_classical_val_array(self, val_array: NDArray[int], debug_str: s
             raise ValueError(f"Negative classical values encountered in {debug_str}")
         if np.any(val_array >= 2**self.bitsize):
             raise ValueError(f"Too-large classical values encountered in {debug_str}")
+
+
+QAnyInt = (QInt, QUInt, BoundedQUInt, QMontgomeryUInt)
+QAnyUInt = (QUInt, BoundedQUInt, QMontgomeryUInt)
+
+
+def _check_uint_fxp_consistent(a: QUInt, b: QFxp) -> bool:
+    """A uint is consistent with a whole or totally fractional unsigned QFxp."""
+    if b.signed:
+        return False
+    return a.num_qubits == b.num_qubits and (b.num_frac == 0 or b.num_int == 0)
+
+
+def check_dtypes_consistent(dtype_a: QDType, dtype_b: QDType, strict: bool = False) -> bool:
+    """Check if two types are consistent given our current definition on consistent types.
+
+    Args:
+        dtype_a: The dtype to check against the reference.
+        dtype_b: The reference dtype.
+        strict: Whether to compare types literally
+
+    Returns:
+        True if the types are consistent.
+    """
+    if dtype_a == dtype_b:
+        return True
+    if strict:
+        return False
+    same_n_qubits = dtype_a.num_qubits == dtype_b.num_qubits
+    if isinstance(dtype_a, QAny) or isinstance(dtype_b, QAny):
+        # QAny -> any dtype and any dtype -> QAny
+        return same_n_qubits
+    elif dtype_a.num_qubits == 1 and same_n_qubits:
```

Comment:
`elif same_n_qubits == 1`?

Reply:
Done.

Reply:
This doesn't work, same_n_qubits is a bool so this may be true even if the types aren't consistent.